File: VB_Classes/Aruco.vb
Imports cv = OpenCvSharp
Imports OpenCvSharp.Aruco.CvAruco

' https://github.com/shimat/opencvsharp_samples/blob/master/SamplesCS/Samples/ArucoSample.cs
Public Class Aruco_Basics : Inherits VB_Algorithm
    Public Sub New()
        desc = "Show how to use the Aruco markers and rotate the image accordingly."
    End Sub
    Public Sub RunVB(src as cv.Mat)
        'Dim tmp = cv.Cv2.ImRead(task.homeDir + "Data/aruco_markers_photo.jpg")
        ''Static detectorParameters = cv.Aruco.DetectorParameters.Create()
        'Static detectorParameters As cv.Aruco.DetectorParameters

        'detectorParameters.CornerRefinementMethod = cv.Aruco.CornerRefineMethod.Subpix
        'detectorParameters.CornerRefinementWinSize = 9
        'detectorParameters.MarkerBorderBits = 8
        'Dim dictionary = cv.Aruco.CvAruco.GetPredefinedDictionary(cv.Aruco.PredefinedDictionaryName.Dict4X4_1000)


        'Dim corners()() As cv.Point2f
        'Dim ids() As Integer
        'Dim rejectedPoints()() As cv.Point2f
        'cv.Aruco.CvAruco.DetectMarkers(tmp, dictionary, corners, ids, detectorParameters, rejectedPoints)
        'dst2 = src.Clone
        'For i = 0 To corners.Count - 1
        '    For Each pt In corners(i)
        '        dst2.Circle(pt, task.dotSize, cv.Scalar.Yellow, -1, task.lineType)
        '    Next
        'Next
        setTrueText("With the OpenCVSharp 4.7, the support for Aruco changed a lot and is not working." + vbCrLf +
                    "Both the C# and the VB.Net versions needed to be commented out.")
    End Sub
End Class




Public Class Aruco_Test : Inherits VB_Algorithm
    Dim aruco As New CS_Classes.Aruco_Detect
    Public Sub New()
        labels(2) = "Original Image with marker ID's"
        labels(3) = "Normalized image after WarpPerspective."
        desc = "Testing the Aruco marker detection in C#"
    End Sub
    Public Sub RunVB(src as cv.Mat)
        Dim tmp = cv.Cv2.ImRead(task.homeDir + "Data/aruco_markers_photo.jpg")
        'aruco.RunCS(tmp)
        'dst2 = aruco.detectedMarkers.Resize(src.Size())

        'dst3.SetTo(0)
        'dst3(New cv.Rect(0, 0, dst3.Height, dst3.Height)) = aruco.normalizedImage.Resize(New cv.Size(dst2.Height, dst2.Height))
        setTrueText("With the OpenCVSharp 4.7, the support for Aruco changed a lot and is not working." + vbCrLf +
                    "Both the C# and the VB.Net versions needed to be commented out.")
    End Sub
End Class

File: CS_Classes/Aruco.cs
using OpenCvSharp.Aruco;
using System;
using System.Collections.Generic;
using cv = OpenCvSharp;

// https://github.com/shimat/opencvsharp_samples/blob/master/SamplesCS/Samples/ArucoSample.cs
namespace CS_Classes
{
    public class Aruco_Detect
    {
        public cv.Mat detectedMarkers;
        public cv.Mat normalizedImage;
        public void RunCS(cv.Mat src)
        {
            // The locations of the markers in the image at FilePath.Image.Aruco.
            //const int upperLeftMarkerId = 160;
            //const int upperRightMarkerId = 268;
            //const int lowerRightMarkerId = 176;
            //const int lowerLeftMarkerId = 168;

            //var detectorParameters = Aruco_Detect();
            //detectorParameters.CornerRefinementMethod = CornerRefineMethod.Subpix;
            //detectorParameters.CornerRefinementWinSize = 9;

            //var dictionary = CvAruco.GetPredefinedDictionary(PredefinedDictionaryName.Dict4X4_1000);

            //CvAruco.DetectMarkers(src, dictionary, out var corners, out var ids, detectorParameters, out var rejectedPoints);
            //detectedMarkers = src.Clone();
            //CvAruco.DrawDetectedMarkers(detectedMarkers, corners, ids, cv.Scalar.White);

            //// Find the index of the four markers in the ids array. We'll use this same index into the
            //// corners array to find the corners of each marker.
            //var upperLeftCornerIndex = Array.FindIndex(ids, id => id == upperLeftMarkerId);
            //var upperRightCornerIndex = Array.FindIndex(ids, id => id == upperRightMarkerId);
            //var lowerRightCornerIndex = Array.FindIndex(ids, id => id == lowerRightMarkerId);
            //var lowerLeftCornerIndex = Array.FindIndex(ids, id => id == lowerLeftMarkerId);

            //// Make sure we found all four markers.
            //if (upperLeftCornerIndex < 0 || upperRightCornerIndex < 0 || lowerRightCornerIndex < 0 || lowerLeftCornerIndex < 0)
            //{
            //    return;
            //}

            //// Marker corners are stored clockwise beginning with the upper-left corner.
            //// Get the first (upper-left) corner of the upper-left marker.
            //var upperLeftPixel = corners[upperLeftCornerIndex][0];
            //// Get the second (upper-right) corner of the upper-right marker.
            //var upperRightPixel = corners[upperRightCornerIndex][1];
            //// Get the third (lower-right) corner of the lower-right marker.
            //var lowerRightPixel = corners[lowerRightCornerIndex][2];
            //// Get the fourth (lower-left) corner of the lower-left marker.
            //var lowerLeftPixel = corners[lowerLeftCornerIndex][3];

            //// Create coordinates for passing to GetPerspectiveTransform
            //var sourceCoordinates = new List<cv.Point2f>
            //{
            //    upperLeftPixel, upperRightPixel, lowerRightPixel, lowerLeftPixel
            //};
            //var destinationCoordinates = new List<cv.Point2f>
            //{
            //    new cv.Point2f(0, 0),
            //    new cv.Point2f(1024, 0),
            //    new cv.Point2f(1024, 1024),
            //    new cv.Point2f(0, 1024),
            //};

            //var transform = cv.Cv2.GetPerspectiveTransform(sourceCoordinates, destinationCoordinates);
            //normalizedImage = new cv.Mat();
            //cv.Cv2.WarpPerspective(src, normalizedImage, transform, new cv.Size(1024, 1024));
        }
    }
}
